### PR TITLE
TSL: Introduce `viewportSafeUV()`

### PR DESCRIPTION
**Pull request by sunag**
Related issue: https://github.com/mrdoob/three.js/pull/28761#issuecomment-2259154805

**Description**

`viewportSafeUV()` is a safe uv function for refraction purposes

### CODE REVIEW

**Comment by mrdoob** [comment]
Looking better!

I'm still seeing some artifacts though:

https://github.com/user-attachments/assets/9449f0a8-062f-4015-8b8c-2e727e2878b6

Notice the part that loses refraction behind the ball:

Screenshot 2024-07-31 at 2 05 51 PM

**Comment by sunag** [comment]
I'm not sure if it will be easy to solve some of these details, maybe it will need a spatial and temporal approach. It is possible to see some of these flaws in AAA games too, since they use similar approaches. It could be less noticeable by reducing the strength of normal too.

We have to consider that the materials currently with `transmission` in `WebGLRenderer` should suffer from these same problems as I show in the example below.

Not that this should be good news, but maybe something our users aren't caring about as much, I think.

```js
// Using WebGLRenderer
const planeRefractor = new THREE.Mesh( planeGeo, new THREE.MeshPhysicalMaterial() );
planeRefractor.material.transparent = true;
planeRefractor.material.transmission = 1;
planeRefractor.material.roughness = .5;
```

https://github.com/user-attachments/assets/29d3edac-2d05-4146-a850-79f289b5be9a

**Comment by mrdoob** [comment]
I see I see...

I guess we just need to update the example so it doesn't show the artifact.
Moving the camera a bit hides it already 😇

File: examples/webgpu_refraction.html [hunk 1]
```diff
@@ -126,7 +126,7 @@
 
 				// renderer
 
-				renderer = new THREE.WebGPURenderer( { antialias: true } );
+				renderer = new THREE.WebGPURenderer( /*{ antialias: true }*/ );
```
Comment (wcandillon) [inline examples/webgpu_refraction.html:129]: @sunag I have to disable also antialias for the example to work as well. Can you provide me context on this?

Comment (wcandillon) [reply examples/webgpu_refraction.html:129]: What I did is set the antiAlias to false like here and handle the pixel density manually in the ctx:
```tsx

    const context = ref.current!.getContext("webgpu")!;
    const canvas = context.canvas;
    canvas.width = canvas.clientWidth * PixelRatio.get();
    canvas.height = canvas.clientHeight * PixelRatio.get();
```

let me know if you have any thoughts on this.

**Pull request merged**